### PR TITLE
Introduce a BuildItem in DevKafkaService to add additional config

### LINKED ISSUE #25094: Introduce a BuildItem in DevKafkaService to add addtional configs
### Description

We want to set ```camel.component.kafka.brokers``` with the value of ```KAFKA_BOOTSTRAP_SERVERS``` if DvKafkaService is enabled. Currently we use the following [codes](https://github.com/apache/camel-quarkus/blob/main/extensions/kafka/deployment/src/main/java/org/apache/camel/quarkus/component/kafka/deployment/KafkaProcessor.java#L66-L81)

```java
 @BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class)
    public void configureKafkaComponentForDevServices(
            DevServicesLauncherConfigResultBuildItem devServiceResult,
            KafkaBuildTimeConfig kafkaBuildTimeConfig,
            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfig) {

        Config config = ConfigProvider.getConfig();
        Optional<String> brokers = config.getOptionalValue(CAMEL_KAFKA_BROKERS, String.class);

        if (brokers.isEmpty() && kafkaBuildTimeConfig.devservices.enabled.orElse(true)) {
            String kafkaBootstrapServers = devServiceResult.getConfig().get(KAFKA_BOOTSTRAP_SERVERS);
            if (kafkaBootstrapServers != null) {
                runTimeConfig.produce(new RunTimeConfigurationDefaultBuildItem(CAMEL_KAFKA_BROKERS, kafkaBootstrapServers));
            }
        }
    }
```
It works in JVM mode but not Native mode. In native test, we expect to launch the application  but there is no ``` -Dcamel.component.kafka.brokers=OUTSIDE://localhost:50172``` in the args. 

It seems that ```RunTimeConfigurationDefaultBuildItem``` will not appear in the native launch args. Is it right? or is there any other possible solution? 

There is a workaround to add this property in the application.properties like
```properties
%prod.camel.component.kafka.brokers=${kafka.bootstrap.servers}
```

So I'd like to introduce a BuildItem to add addtional configs when DevServiceKafka returns ```DevServicesResultBuildItem``` which can add an addtional config. I'm not sure this is a common case or it is overkill.

### Implementation ideas

I'd like to introduce
```java
package io.quarkus.kafka.client.deployment;

import io.quarkus.builder.item.MultiBuildItem;

public final class DevServiceKafkaAdditionalBuildItem extends MultiBuildItem {
    private final String config;

    public DevServiceKafkaAdditionalBuildItem(String config) {
        this.config = config;
    }

    public String getConfig() {
        return config;
    }
}
```

and in ```DevServicesKafkaProcessor.java``` to add this addtional config when returning ```DevServicesResultBuildItem```
```java
private Map<String, String> getConfigs(String address, List<DevServiceKafkaAdditionalBuildItem> additions) {
        HashMap<String, String> configs = new HashMap<>();
        configs.put(KAFKA_BOOTSTRAP_SERVERS, address);
        additions.forEach(c -> {
            configs.put(c.getConfig(), address);
        });

        return configs;
    }
```

So I can produce a ```DevServiceKafkaAdditionalBuildItem``` in quarkus-camel-kafka
```java
@BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class)
    public void configureKafkaComponentForDevServices(
            KafkaBuildTimeConfig kafkaBuildTimeConfig,
            BuildProducer<DevServiceKafkaAdditionalBuildItem> additions) {

        Config config = ConfigProvider.getConfig();
        Optional<String> brokers = config.getOptionalValue(CAMEL_KAFKA_BROKERS, String.class);

        if (brokers.isEmpty() && kafkaBuildTimeConfig.devservices.enabled.orElse(true)) {
            additions.produce(new DevServiceKafkaAdditionalBuildItem(CAMEL_KAFKA_BROKERS));
        }
```

### PR DESCRIPTION
**Pull request by zhfeng**
Fix #25094

### CODE REVIEW

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServiceKafkaAdditionalBuildItem.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package io.quarkus.kafka.client.deployment;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class DevServiceKafkaAdditionalBuildItem extends MultiBuildItem {
+    private final String config;
```
Comment (gastaldi) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServiceKafkaAdditionalBuildItem.java:6]: Can you give an example on what value would this config attribute have?

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 2]
```diff
@@ -255,18 +257,28 @@ private RunningDevService startKafka(KafkaDevServiceCfg config,
...
-                        KAFKA_BOOTSTRAP_SERVERS, containerAddress.getUrl()))
+                        getConfigs(containerAddress.getUrl(), additions)))
                 .orElseGet(defaultKafkaBrokerSupplier);
     }
 
+    private Map<String, String> getConfigs(String address, List<DevServiceKafkaAdditionalBuildItem> additions) {
+        HashMap<String, String> configs = new HashMap<>();
+        configs.put(KAFKA_BOOTSTRAP_SERVERS, address);
+        additions.forEach(c -> {
+            configs.put(c.getConfig(), address);
```
Comment (gastaldi) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:276]: The `BuildItem` is used exclusively to set the Kafka's server address as a value for other keys?

Comment (gastaldi) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:276]: Looking at the original issue, it seems to be `camel.component.kafka.brokers`

Comment (zhfeng) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:276]: yeah, we use it in  ```camel.component.kafka.brokers``` see https://github.com/apache/camel-quarkus/issues/3742

Comment (zhfeng) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:276]: Also you can check the detail about the motivation of these changes in https://github.com/quarkusio/quarkus/issues/25094

Comment (gastaldi) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:276]: I see, I am just concerned that the class name `DevServiceKafkaAdditionalBuildItem` may be too generic for this reason

Comment (zhfeng) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:276]: OK, it is hard to name it :) any thought ?

**Review by ozangunalp — COMMENTED** [review]
I don't think that this is the right solution for this problem.
We need to understand why `DevServicesLauncherConfigResultBuildItem` doesn't contain the dev service config in native mode. 
[The code here](https://github.com/apache/camel-quarkus/blob/main/extensions/kafka/deployment/src/main/java/org/apache/camel/quarkus/component/kafka/deployment/KafkaProcessor.java#L66-L81) seems to be the right way of getting the dev service config.

**Comment by zhfeng** [comment]
But in camel-quarkus-kafka, the above build step seems not run at all in native mode. I'm not sure it is because we produce ```RunTimeConfigurationDefaultBuildItem``` which do not be consumered in native?

**Comment by ozangunalp** [comment]
@zhfeng In integration tests, the application is built in "normal" mode, without dev services or test mode augmentation.
The availability of dev services for native integration tests is somewhat _simulated_ with a separate build and injected to the application config, as you see it on the logs `-Dkafka.bootstrap.servers=...`

That's why contrary to what I wrote before, the previous code is not the right way of doing this. 

In hindsight, I don't completely understand the need to have a `camel.component.kafka.brokers` default config only for the tests.
I think it is better to handle this case inside the test case. 
You can check out `io.quarkus.test.common.DevServicesContext` which lets accessing dev service properties from tests, including integration tests.

Maybe @geoand would have another take on this?

Ref #25094

**Comment by geoand** [comment]
> In hindsight, I don't completely understand the need to have a camel.component.kafka.brokers default config only for the tests.

I don't undertand this requirement at all either :)

@zhfeng can you describe the end goal of what you are trying to do?

**Comment by zhfeng** [comment]
@ozangunalp camel-quarkus-kafka extension reads ```camel.component.kafka.brokers``` and autowire it into the camel-kafka during the camel-quarkus based application bootstrap.

@gastaldi I want to get the ```camel.component.kafka.broker``` with the same value of ```kafka.bootstrap.servers``` automatically both in jvm and native mode. And in native, just like ```-Dcamel.component.kafka.brokers=...``` in args when launching  the native test. Currently there is a workaround to set it in application.properties
```properties
%prod.camel.component.kafka.brokers=${kafka.bootstrap.servers}
```

@jamesnetherton do you have any input here?

**Comment by zhfeng** [comment]
@ozangunalp I find some similar codes with [HibernateOrmProcessor.java](https://github.com/quarkusio/quarkus/blob/main/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L263-L311) which produce ```RunTimeConfigurationDefaultBuildItem``` to set some properties on Dev Services managed database.

So I wonder this is only availiable in JVM  but not Native testing?

**Comment by ozangunalp** [comment]
Same for hibernate, that runtime default config would not be set on the application under test during integration tests.
That is why you would actually need `quarkus.hibernate-orm.database.generation=drop-and-create` in properties file for those ITs to work.

**Comment by zhfeng** [comment]
Thanks for explanation. So I think it could be much more reasonable to introduce such a BuilidItem in DevServiceKafka which could be very helpful for other extensions in IT native testing.

**Comment by ozangunalp** [comment]
I couldn't find where does the `camel.component.kafka.brokers` property is defined. If it is a safe assumption you can write `camel.component.kafka.brokers=${kafka.bootstrap.servers}` as a default configuration property in the camel extension.

**Comment by zhfeng** [comment]
yeah, we have an example https://github.com/apache/camel-quarkus-examples/tree/main/kafka. We only want ```camel.component.kafka.brokers=${kafka.bootstrap.servers}``` in ```Dev``` or ```Test``` profile when DevKafkaService is enabled.  That's the reason we introudce a [BuildStep](https://github.com/apache/camel-quarkus/blob/main/extensions/kafka/deployment/src/main/java/org/apache/camel/quarkus/component/kafka/deployment/KafkaProcessor.java#L66-L81)  in camel-quarkus-kafka extension.

Since the native IT test is running in ```prod``` profile, https://github.com/apache/camel-quarkus-examples/pull/86 could be a good way.

**Comment by cescoffier** [comment]
Shouldn't this be done by the camel extension? 
Our dev service for Kafka can emit a build item with the configuration if needed, but that's all we can do. 

As highlighted before, having

```
camel.component.kafka.brokers=${kafka.bootstrap.servers} 
 ```
 
 is the right way to do it Today.

**Comment by zhfeng** [comment]
Thanks @cescoffier ! As I understand, the purpose of dev services is avoiding or lessing the configurations when running the tests. That was the motivation we did in camel-quarkus-kafka extension before. So I think it is only possible to achieve this target both in JVM and Native mode by introducing  a BuildItem to produce an additional configuration.

**Comment by cescoffier** [comment]
@zhfeng I do not follow. First, dev services are not about reducing the configuration, but about running infrastructure services automatically. 
Then, why can't you in the camel-quarkus-kafka define the property as mentioned above?

**Comment by zhfeng** [comment]
@cescoffier yeah, we can define the property but ```kafka.bootstrap.servers``` is only availiable when kafka dev service is enabled. We have to run the tests in different environments (dev services enabled and disabled).

**Comment by cescoffier** [comment]
This property is set by the dev service, but is used every time. So in prod mode, the user would have to set it (or use SBO).

**Comment by zhfeng** [comment]
Sorry, what is ```SBO``` ?

**Comment by cescoffier** [comment]
Service Binding Operator

**Pull request closed without merging**